cmd-ref: plots: add templates docs (#3697)
iterative/dvc.org: content/docs/command-reference/plots/index.md

## Plot templates (data series only)

> DVC uses [Vega-Lite](https://vega.github.io/vega-lite/) JSON specifications to
> create visualizations out of users data. Pre-defined specifications are called
> templates.

Reviewer: ```suggestion
DVC uses [Vega-Lite](https://vega.github.io/vega-lite/) JSON templates to
create visualizations out of users' data.
```

Optional suggestion. I don't know that we need to explain what a template means.

Reviewer 2: I'd keep "specifications" since it's the term used in the linked page (explains what templates are but very implicitly). I'd also pull a sentence from the following paragraph here:

```suggestion
DVC uses [Vega-Lite](https://vega.github.io/vega-lite/) JSON specifications
to create plots from user data. Built-in _plot templates_ are stored within the
[`dvc-render`](https://github.com/iterative/dvc-render) library.
```

> DVC has few built in templates. They are stored within
> [`dvc-render`](https://github.com/iterative/dvc-render) library. The default one

Reviewer: ```suggestion
DVC has few built in templates. They are stored within the
[`dvc-render`](https://github.com/iterative/dvc-render) library. The default one
```

> [`dvc-render`](https://github.com/iterative/dvc-render) library. The default one

Reviewer: Optional suggestion: Can we drop the reference to dvc-render?

Reviewer 2: I did remove it for now but do we plan to add a bunch? If so maybe document then in https://github.com/iterative/dvc-render#readme and just list a few here + link there.

Author: I think there should be a note somewhere about it - if we want people to contribute templates, we should include the link somewhere in `index`.

Reviewer 2: > if we want people to contribute templates

If that's the intention then please do link to the repo somewhere.

is called `linear`. It can be changed with the `--template` (`-t`) option of
`dvc plots show` and `dvc plots diff`. The argument provided to `--template`

> option can be either a built-in template name or path to a JSON specification.

Reviewer: ```suggestion
can be either a built-in template name or path to a JSON specification.
```

> DVC has a few built-in templates. They are stored within
> [`dvc-render`](https://github.com/iterative/dvc-render) library. The default one
> is called `linear`. It can be changed with the `--template` (`-t`) option of
> `dvc plots show` and `dvc plots diff`. The argument provided to `--template`
> can be either a built-in template name or path to a JSON specification.
> For templates stored in the `.dvc/plots/` directory, the path and the json
> extension are not required: user can specify only the base name e.g.
> `--template scatter`.

Reviewer 2: ```suggestion
The `linear` template is the default. It can be changed with the `--template`
(`-t`) option of `dvc plots show` and `dvc plots diff`. The argument provided to
`--template` can be a (built-in) template name or a path to a
[custom template](https://dvc.org/doc/command-reference/plots/templates).

<admon type="tip">

For templates stored in `.dvc/plots`, the path and the json
extension are not required: user can specify only the base name e.g.
`--template scatter`.

</admon>
```

Reviewer: Which is the default behavior. It's a bit awkward in part to not break behavior for older repos that already have templates there.

DVC has the following built-in plot templates:

> - `linear` - linear plot, used by default

Reviewer 2: ```suggestion
- `linear` - linear plot (default)
```

- `scatter` - scatter plot
- `smooth` - linear plot with LOESS smoothing, see
  [example](/doc/command-reference/plots#example-smooth-plot)
- `confusion` - confusion matrix, see
  [example](/doc/command-reference/plots#example-confusion-matrix)
- `confusion_normalized` - confusion matrix with values normalized to <0, 1>
  range
- `simple` - simplest linear template, good for learning how to work with

>   [`templates`](/doc/command-reference/plots/templates) command

Reviewer: Isn't it more useful as a baseline for creating custom templates?

Reviewer 2: What's the difference with `linear`?

Author: Main difference is lack of interactivity in simple. It also has like 4 times less lines, so its easier to comprehend. Moved up.

File targets given to `dvc plots show` and `dvc plots diff` are treated as
separate data series, each to be injected into a template file. There are two
important fields that DVC adds to the plot data:

- `step` - zero-based counter for the data rows/values. In many cases it
  corresponds to a machine learning training epoch number.

- `rev` - This field helps distinguish between data sourced from different

>   revisions, files or columns.

Reviewer 2: Should we hide this in a `<details>` block? Doesn't seem like something 90% of users need to know.

Note that in the case of CSV/TSV metrics files, column names from the table
header (first row) are equivalent to field names.

> ### Custom templates
> 

Reviewer 2: Prob not needed if we link to `plot templates` per one of my suggestions above.
```suggestion
```

Author: Not sure how to proceed, didn't remove it because I followed @dberenbaum suggestion from 
https://github.com/iterative/dvc.org/pull/3697#discussion_r908790245

Reviewer 2: Np let's leave it for now. Thanks

> Users can create their own templates. Refer to
> [`templates`](/doc/command-reference/plots/templates) command for more
> information.
> 

Reviewer 2: Same
```suggestion
```

## Plot templates (data series only)

> DVC uses [Vega-Lite](https://vega.github.io/vega-lite/) JSON specifications to
> create visualizations out of users' data. Pre-defined specifications are called
> templates.

Reviewer 2: Repeat of https://github.com/iterative/dvc.org/pull/3697#discussion_r910466305 + https://github.com/iterative/dvc.org/pull/3697#pullrequestreview-1022147287
```suggestion
DVC uses [Vega-Lite](https://vega.github.io/vega-lite/) JSON specifications
to create plots from user data. A set of built-in _plot templates_ are included.
```

The `linear` template is the default. It can be changed with the `--template`
(`-t`) option of `dvc plots show` and `dvc plots diff`. The argument provided to
`--template` can be a (built-in) template name or a path to a

> [custom template](https://dvc.org/doc/command-reference/plots/templates).

Reviewer 2: ```suggestion
[custom template].
```

<admon type="tip">
For templates stored in `.dvc/plots`, the path and the json
extension are not required: user can specify only the base name e.g.
`--template scatter`.
</admon>

DVC has the following built-in plot templates:

> - `linear` - linear plot (default)
> - `simple` - simplest linear template, good base for creating custom template
>   and learning how to work with
>   [`templates`](/doc/command-reference/plots/templates) command

Reviewer 2: ```suggestion
- `linear` - basic linear plot including cursor interactivity (default)
- `simple` - simplest linear template (not interactive); Good base to create
  [custom templates].
```

- `scatter` - scatter plot
- `smooth` - linear plot with LOESS smoothing, see
  [example](/doc/command-reference/plots#example-smooth-plot)

> - `confusion` - confusion matrix, see
>   [example](/doc/command-reference/plots#example-confusion-matrix)

Reviewer 2: ```suggestion
- `confusion` - confusion matrix, see
  [example](/doc/command-reference/plots#example-confusion-matrix)

[custom template]: https://dvc.org/doc/command-reference/plots/templates

```

## Plot templates (data series only)

DVC uses [Vega-Lite](https://vega.github.io/vega-lite/) JSON specifications
to create plots from user data. A set of built-in _plot templates_ are included.

The `linear` template is the default. It can be changed with the `--template`
(`-t`) option of `dvc plots show` and `dvc plots diff`. The argument provided to
`--template` can be a (built-in) template name or a path to a
[custom template].

<admon type="tip">

For templates stored in `.dvc/plots` (default location for custom templates),
the path and the json extension are not required: you can specify only the
base name, e.g. `--template scatter`.

</admon>

DVC has the following built-in plot templates:

- `linear` - basic linear plot including cursor interactivity (default)
- `simple` - simplest linear template (not interactive); Good base to create

>   [custom templates].

Reviewer 2: Not for this PR: Should we talk more about this interactivity? I'm not even sure what it is and it's not mentioned much (at all actually) in other `plots` refs.

Author: I don't think so - interactivity is just to make user experience better, but its nothing major:
[simple example](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JACyUAVhDYA7EABoQAEySYUqUMSSCGcCGgDaoJFCgMATqYCeaEAAZKAdjEAOAJz2ATGM-2AbADMrgCMXgCsiiAAZjSCcLJICHB2wqScUDpIFhAEpuZWUNaUmBDEURZw5egA7mwWANYQOKYpShCYcDh29lGYNMkdSd3oIf7hTl5hU-5eTpEAvgomZpY2PZSuk06BIfZOk57+HlGx8YnJqTTp0Fk5eauFxaXlSpXVIHWNza1RHV12EJ9Ab6dQIEYgMYTKZeGZeVy9JYrArrdCOdxhMRiXYuQLhQIE-ynOIJJIpdBpUS3QTZXL5NZFEplCpVOxfJotKBtED-CFeYGDMEQqGTaZjeGLZYgelPDbuYLhMThVz+ELhEKhbyRJRnUmXCnXKmZGn3GU2JmvEDvNn1Dm-dqdCGBAWg4aA8ai2Hik5I6WPVEOSjuNyBNz2cL4jVKwL8nUki7kkCUjJ3On+xkvFkfdk-Ll-R12MQuobg93QsWzeUgX1mopy8LuULzdxzcL2dyzZ1x85kq43Y20h4ojPMt6s2q23Pc3l2bUgfqCt2jD0wuHuJzVqW12xooP+Ft4+weEJOffzDfdvWJ5PUwfbi1Zm3fTnTgvoIlKBeu0vL8teoL2ECNbpjugYdi4riHtiTgYg2sYxPGvYGv2qZDgyzyjla46fJOL75gC6AXvOIIlsKK4VoE4absi6FypMXiBJMrjwv4QSwq4xI9vqSaGimJppsOGGWtaE7PvaPJviAHGfiRQplp6MyUe41YALpLMRmDxHYUQ1DQyj0GglH2Eo8A0GQWCGfYxkgMgjRoKAmDWDg140LIKTqQkUBsMorlkPZIAAB7+bEcCCModgzp+TmJgAjgwSCyP06j9KQwKaYmM7qTuoAhWFdjbn00V2HFCVJWohppVp6AFe0xrkqAABecAWGwaDRDoEBwAs6leYI9TBTQoXheg1pRc5diyGwCCuTo1bdUAA)

[interactive](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JACyUAVhDYA7EABoQAEySYUqUMSSCGcCGgDaoJFCgMATqYCeaEAAZKAdjEAOAJz2ATGM-2AbADMrgCMXgCsiiAAZjSCcLJICHB2wqScUDpIFhAEpuZWUNaUmBDEURZw5egA7mwWANYQOKYpShCYcDh29lGYNMkdSd3oIf7hTl5hU-5eTpEAvgomZpY2PZSuk06BIfZOk57+HlGx8YnJqTTp0Fk5eauFxaXlSpXVIHWNza1RHV12EJ9Ab6dQIEYgMYTKZeGZeVy9JYrArrdCOdxhMRiXYuQLhQIE-ynOIJJIpdBpUS3QTZXL5NZFEplCpVOxfJotKBtED-CFeYGDMEQqGTaZjeGLZYgelPDbuYLhMThVz+ELhEKhbyRJRnUmXCnXKmZGn3GU2JmvEDvNn1Dm-dqdCGBAWg4aA8ai2Hik5I6WPVEOSjuNyBNz2cL4jVKwL8nUki7kkCUjJ3On+xkvFkfdk-Ll-R12MQuobg93QsWzeUgX1mopy8LuULzdxzcL2dyzZ1x85kq43Y20h4ojPMt6s2q23Pc3l2bUgfqCt2jD0wuHuJzVqW12xooP+Ft4+weEJOffzDfdvWJ5PUwfbi1Zm3fTnTgvoIlKBeu0vL8teoL2ECNbpjugYdi4riHtiTgYg2sYxPGvYGv2qZDgyzyjla46fJOL75gC6AXvOIIlsKK4VoE4absi6FypMXiBJMrjwv4QSwq4xI9vqSaGimJppsOGGWtaE7PvaPJviAHGfiRQplp6MyUe41YALpLMRmDxHYUQ1DQyj0GglH2Eo8A0GQWCGfYxlJkg1hwBYRigAkUBsMoNCyGQaCgAAHl5CFwIIyh2DOn7WDgiYAI4MEgsj9Oo-SkMCmmJjO6k7qAsQBUF6Dbn0YWRdFsV0GohpJVpOUgX8xrkqAABe9lsGg0Q6BAcALOpLmCPUfmZYFdjWqF4V2LIbAIO5OjVupNJ2Q5qDGCAyCNFcsgpL6rXxFA-RyH5NJMAFfmYPlwXuWQWlKCt2SgmgmAWHoSjbegCBsAwrVsKQDlKM5rknQYc0gL5KmfeCh3DXI3JQPE2R2E9L1wM9mCTUoi0NHYOBsO5COfbILluR5flsJydDpSALmyG5W3yJoPIBXAm3sJTNl7YIUTaLo5IhOprN6Gg9jtQsalSjdMUQNE9QII5CGCJ0s2gOttMU6kSBM5NgM2TNEvIwdR3oLdZ0k2wXWzSAZBWLY6lfbjnlU75VO9dlEkAoNBUxXFJWJXzUoWydB1wN5CNU4dQ3oFFLvFQl3J23YuXqTbGU0FlwUFk7dgh0V8WlWlPXx31FWCXlQcgKnrvh4jav2Rr2QowH2vzr7mPSsIZAM-E0T18oNvhEoyg7vw4Tm9j3141TnXddXBcjWNiTM3GCc66yfNqYvCxAA) - hover with mouse over the plot

Reviewer 2: OK. Are most templates interactive though? (`layer/mark` element in the spec, I think)

We're only mentioning that these 2 are/are not. Minor indeed
